packages/cli/src/lib/vite/worker-entry.ts

```diff
@@ -0,0 +1,198 @@
+/**
+ * All the code in this file is executed in workerd. This file
+ * is compiled at build time (tsup) to be transpiled to JS, and
+ * then loaded as string in a workerd instance at runtime as
+ * the worker entrypoint. It then requests modules to Vite
+ * and evaluates them to run the app's backend code.
+ */
+
+import {
+  ViteRuntime,
+  ssrModuleExportsKey,
+  type ViteRuntimeModuleContext,
+} from 'vite/runtime';
+import type {HMRPayload} from 'vite';
+import type {Response} from 'miniflare';
+
+export interface ViteEnv {
+  __VITE_ROOT: string;
+  __VITE_HMR_URL: string;
+  __VITE_FETCH_MODULE_PATHNAME: string;
+  __VITE_RUNTIME_EXECUTE_URL: string;
+  __VITE_WARMUP_PATHNAME: string;
+  __VITE_SETUP_ENV: (request: Request) => void;
+  // Ref: https://github.com/cloudflare/workerd/blob/main/src/workerd/api/unsafe.h
+  __VITE_UNSAFE_EVAL: {
+    eval(code: string, name?: string): Function;
+    newFunction(code: string, name?: string, ...args: string[]): Function;
+    newAsyncFunction(code: string, name?: string, ...args: string[]): Function;
+  };
+}
+
+const O2_PREFIX = '[o2:runtime]';
+
+export default {
+  /**
+   * Worker entry module that wraps the user app's entry module.
+   */
+  async fetch(request: Request, env: ViteEnv, ctx: any) {
+    env.__VITE_SETUP_ENV(request);
+    const url = new URL(request.url);
+
+    // Fetch the app's entry module and cache it. E.g. `<root>/server.ts`
+    const module = await fetchEntryModule(url, env);
+
+    // Return early for warmup requests after loading the entry module.
+    if (url.pathname === env.__VITE_WARMUP_PATHNAME) {
+      return new globalThis.Response(null);
+    }
+
+    // Execute the user app's entry module.
+    return module.default.fetch(request, createUserEnv(env), ctx);
```

Comment:
Inception 😵‍💫

Reply:
Hahah yeah! And I think this could be nice to control the `globalThis` in development environment from here. Also, I think we could import `AsyncLocalStorage` at this point with `nodejs_compat` but still disable it somehow for the user code 🤔